pkg/controller/master/upgrade/version_syncer.go

```diff
@@ -174,21 +178,24 @@ func (s *versionSyncer) syncVersions(resp CheckUpgradeResponse, currentVersion s
 			}
 		}
 	}
+
 	return nil
 }
 
 // cleanupVersions remove version resources that's can't be upgraded to anymore
-func (s *versionSyncer) cleanupVersions(currentVersion string) error {
+func (s *versionSyncer) cleanupVersions(currentVersion string, remoteVersions []Version) error {
 	versions, err := s.versionClient.List(s.namespace, metav1.ListOptions{})
 	if err != nil {
 		return err
 	}
 
-	for i := range versions.Items {
-		v := versions.Items[i]
-		if !canUpgrade(currentVersion, &v) {
-			if err := s.versionClient.Delete(v.Namespace, v.Name, &metav1.DeleteOptions{}); err != nil {
-				return err
+	for _, version := range remoteVersions {
+		for i := range versions.Items {
```

Comment:
the second for loop looks a bit redundant:

the new `canUpgrade` mainly check the `current` and `remote`; mostly `!canUpgrade` will let second loop run and delete those item repeatly

Reply:
multiple versions may be possible for a current version based on the remoteVersions returned from upgrade responder. 

We need to run the inner loop to reconcile all local version CRD's with what is returned by upgrade responder. 

With this change we can rollback a version by bumping up the remoteVersion in upgrade responder to a really high semver say 1.99.x which would never be satisfied and trigger clean up version CRD's in clusters when they fetch the latest remote versions, there by preventing users from upgrading to the version.

Reply:
Suppose a local version object is removed due to `!canUpgrade` in the first iteration. The removed version object will be iterated again with the following comparison round (because it's a nested loop and we don't remove the element from the list). The same local version object might fail the test again, causing the subsequent version deletion in a "version not found" error.

Reply:
thanks for the feedback, I have fined tuned the inner loop to drop deleted version objects and also ignore errors if a version object is not found during deletion